lib/Dialect/TritonGPU/Transforms/OptimizeThreadLocality.cpp

```diff
@@ -0,0 +1,278 @@
+#include "mlir/Pass/Pass.h"
+#include "mlir/Pass/PassManager.h"
+#include "mlir/Transforms/Passes.h"
+#include "triton/Dialect/TritonGPU/IR/Dialect.h"
+#include "triton/Dialect/TritonGPU/Transforms/Passes.h"
+#include "triton/Dialect/TritonGPU/Transforms/Utility.h"
+#include <memory>
+#define GEN_PASS_CLASSES
+#include "triton/Dialect/TritonGPU/Transforms/Passes.h.inc"
+
+using namespace mlir;
+
+class TritonGPUOptimizeThreadLocalityPass
+    : public TritonGPUOptimizeThreadLocalityBase<
+          TritonGPUOptimizeThreadLocalityPass> {
+  void runOnOperation() override {
+    ModuleOp mod = getOperation();
+    DenseSet<triton::ReduceOp> reduceOps;
+    mod.walk([&](triton::ReduceOp reduce) -> void {
+      auto srcType = reduce.getOperands()[0].getType().cast<RankedTensorType>();
+      auto rank = srcType.getShape().size();
+      auto srcEncoding = srcType.getEncoding();
+      if (!(srcEncoding.isa<triton::gpu::BlockedEncodingAttr>() && rank == 2))
+        return;
+      if (!reduce->hasOneUse())
+        return;
+      Operation *user = (reduce->getUses().begin())->getOwner();
+      if (!user->hasOneUse())
+        return;
+      OpOperand &yieldOpOperand = *(user->getUses().begin());
+      auto yieldOp = dyn_cast<scf::YieldOp>(yieldOpOperand.getOwner());
+      if (!yieldOp)
+        return;
+      auto operandNumber = yieldOpOperand.getOperandNumber();
+      Block *block = reduce->getBlock();
+      Operation *parentOp = block->getParentOp();
+      auto forOp = dyn_cast<scf::ForOp>(parentOp);
+      if (!forOp)
+        return;
+      auto argNum = yieldOpOperand.getOperandNumber();
+      auto oldAccum = forOp.getInitArgs()[argNum];
+      auto cstOp = dyn_cast<arith::ConstantOp>(oldAccum.getDefiningOp());
+      if (!cstOp)
+        return;
+      auto oldAccumValue = cstOp.getValue();
+      auto denseAttr = oldAccumValue.dyn_cast<DenseFPElementsAttr>();
+      if (!(denseAttr.isSplat() && denseAttr.getSplatValue<APFloat>().isZero()))
```

Comment:
Need to support non-zero splat by initializing the new accumulator with a neutral value (based on the reduction) then incorporating the splat value to the result (after the loop)